pkg/library/flatten/merge.go

```diff
@@ -193,6 +214,52 @@ func mergeContainerContributions(flattenedSpec *dw.DevWorkspaceTemplateSpec) err
 	return nil
 }
 
+// Finds a component that is a suitable merge target for container contributions and returns its name.
+// The following rules are followed when finding a merge target:
+//
+// - A container component that has the merge-contribution: true attribute will automatically be selected as a merge target.
+//
+// - A container component that has the merge-contribution: false attribute will be never be selected as a merge target.
+//
+// - Otherwise, the first container component found that was not imported by a plugin or parent devworkspace (i.e. the controller.devfile.io/imported-by attribute is not present)
+// will be selected as a merge target.
+//
+// If no suitable merge target is found, an error is returned.
+func findMergeTarget(flattenedSpec *dw.DevWorkspaceTemplateSpec) (mergeTargetComponentName string, err error) {
+	// First check for explicit merge contributtion attribute
+	for _, component := range flattenedSpec.Components {
+		if component.Container == nil {
+			continue
+		}
+
+		if component.Attributes.Exists(constants.MergeContributionAttribute) {
+			if component.Attributes.GetBoolean(constants.MergeContributionAttribute, nil) {
+				return component.Name, nil
+			}
+		}
+	}
+
+	// Then see if there's a container that can implicitly be selected as a merge target
+	for _, component := range flattenedSpec.Components {
+		if component.Container == nil {
+			continue
+		}
+
+		// Don't select components that opt out as a merge contribution target
+		if component.Attributes.Exists(constants.MergeContributionAttribute) && !component.Attributes.GetBoolean(constants.MergeContributionAttribute, nil) {
+			continue
+		}
+
+		// The target must not have been imported by a plugin or parent.
+		if component.Attributes.Exists(constants.PluginSourceAttribute) {
+			continue
+		}
+		return component.Name, nil
+	}
```

Comment:
This could be simplified into a single pass in a pretty straightforward way:

```go
firstComponent := ""
for _, component := range flattenedSpec.Components {
	if component.Container == nil {
		continue
	}

	if component.Attributes.Exists(constants.MergeContributionAttribute) {
		if component.Attributes.GetBoolean(constants.MergeContributionAttribute, nil) {
			return component.Name, nil
		}
	}

	if component.Attributes.Exists(constants.PluginSourceAttribute) {
		continue
	}

	if firstComponent == "" {
		firstComponent = component.Name
	}
}
```

Reply:
Good call, I added a fixup for this (I included an extra `continue` to prevent accidentally selecting a component that opts out as a merge contribution target)